Monika After Story/game/zz_hangman.rpy

```diff
@@ -496,18 +489,62 @@ label mas_hangman_game_loop:
         m "[hm_hint]"
 
     # main loop for hangman game
-    $ done = False
-    $ win = False
-    $ chances = 6
-    $ guesses = 0
-    $ missed = ""
-    $ avail_letters = list(hm_ltrs_only)
-    $ give_up = False
-
-    if persistent._mas_sensitive_mode:
-        $ avail_letters.remove("?")
-
-    $ dt_color = mas_hmg.WORD_COLOR
+    python:
+        done = False
+        win = False
+        chances = 6
+        guesses = 0
+        last_chance_guesses = 0
+        missed = ""
+        avail_letters = list(hm_ltrs_only)
+        give_up = False
+
+        if persistent._mas_sensitive_mode:
+            avail_letters.remove("?")
+
+        dt_color = mas_hmg.WORD_COLOR
+
+        def getCorrectGuesses(guesses, chances, max_chances=6):
+            """
+            Get the amount of correct guesses, i.e. the amount of guesses 
+            that actually exposed letters in the word.
+
+            IN:
+                guesses - amount of guesses (including incorrect) made
+                    by the player
+                chances - amount of chances the player has
+                max_chances - maximum amount of chances player can
+                    have (defaults to 6)
+
+            RETURNS:
+                Amount of correct guesses.
+            """
+            return guesses + chances - max_chances
+
+        def getLastChanceGuessesRatio(guesses, chances, last_chance_guesses,
+                                      useCorrect=False):
+            """
+            Get the ratio of so-called 'last chance guesses' (correct guesses
+            made on the last chance remaining) to the total amount of guesses
+            (depending on 'useCorrect' parameter, this may be either both 
+            correct and incorrent amount, or just amount of correct guesses.)
+            
+            IN:
+                guesses - amount of guesses (including incorrect) made
+                    by the player
+                chances - amount of chances the player has
+                last_chance_guesses - amount of guesses (including
+                    incorrect) made by the player on their last chance
+                useCorrect - whether to use total amount of only correct
+                    guesses or total amount of any guesses (defaults to False)
+
+            RETURNS:
+                Float in range [0; 1] representing last chance guesses to
+                correct guesses ratio.
+            """
+            return last_chance_guesses / (getCorrectGuesses(guesses, chances) 
+                   if useCorrect else guesses)
```

Comment:
Please don't multiline like this, it's harder to read, use parenthesis to enclose the full return if you need multiple lines, or just keep it to a single line.

```suggestion
            return last_chance_guesses / (getCorrectGuesses(guesses, chances) if useCorrect else guesses)
```